### ISSUE TITLE
Rule proposal: no-unused-async-promise

### ISSUE THREAD

**Issue by jedwards1211** [eslint/eslint#8473]
**Please describe what the rule should do:**

It should generate an error when a statement calls an async function without using the `Promise` returned by the async function.  I've often mistakenly typed `foo()` instead of `await foo()`, causing bugs.

**What category of rule is this? (place an "X" next to just one item)**

[ ] Enforces code style
[X] Warns about a potential error
[ ] Suggests an alternate way of doing something
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

```js
async function foo() {
  ...
}

function bar() {
  foo() // error: unused Promise returned from foo
}

// the following are all valid code under this rule:
async function baz() {
  await foo()
}
function qux() {
  return Promise.all([foo(), foo()])
}
function qlorm() {
  return foo()
}
function gloob() {
  const x = foo() // `x` is unused, but that should be handled by `no-unused-vars`
}
```

**Why should this rule be included in ESLint (instead of a plugin)?**

I'm not 100% sure it should be included in ESLint.  But I figured I would put this up for discussion anyway.

**Comment by not-an-aardvark** [comment]
Thanks for the rule suggestion. I think it would be useful to be able to detect errors like this, but I'm concerned that ESLint wouldn't be very effective at detecting them due to the limits of static analysis. Specifically, ESLint would only be able to tell that `foo` refers to an async function if it's explicitly declared in the same file (and it wouldn't be able to detect async method calls like `object.foo()`.)

Bluebird has a [loosely related runtime check](http://bluebirdjs.com/docs/warning-explanations.html#warning-a-promise-was-created-in-a-handler-but-was-not-returned-from-it) for something like this, although it's detecting a slightly different problem.

**Comment by kaicataldo** [comment]
While I agree this is useful, I don't see how this is possible given the limitations of static analysis. Flow/TypeScript seems like a better tool for catching these kinds of errors.

**Comment by jedwards1211** [comment]
Oh, you're right!  I didn't think about the same file limitation.  Thanks for letting me know!
